service/listeners.go

```diff
@@ -0,0 +1,382 @@
+
+// StreamListener is a network listener for stream-oriented protocols that
+// accepts [transport.StreamConn] connections.
+type StreamListener interface {
+	// Accept waits for and returns the next connection to the listener.
+	AcceptStream() (transport.StreamConn, error)
+
+	// Close closes the listener.
+	// Any blocked Accept operations will be unblocked and return errors. This
+	// stops the current listener from accepting new connections without closing
+	// the underlying socket. Only when the last user of the underlying socket
+	// closes it, do we actually close it.
+	Close() error
+
+	// Addr returns the listener's network address.
+	Addr() net.Addr
+}
+
+type TCPListener struct {
+	ln *net.TCPListener
+}
+
+var _ StreamListener = (*TCPListener)(nil)
+
+func (t *TCPListener) AcceptStream() (transport.StreamConn, error) {
+	return t.ln.AcceptTCP()
+}
+
+func (t *TCPListener) Close() error {
+	return t.ln.Close()
+}
+
+func (t *TCPListener) Addr() net.Addr {
+	return t.ln.Addr()
+}
+
+type OnCloseFunc func() error
+
+type acceptResponse struct {
+	conn transport.StreamConn
+	err  error
+}
+
+type virtualStreamListener struct {
+	mu          sync.Mutex // Mutex to protect access to the channels
+	addr        net.Addr
+	acceptCh    <-chan acceptResponse
+	closeCh     chan struct{}
+	onCloseFunc OnCloseFunc
+}
+
+var _ StreamListener = (*virtualStreamListener)(nil)
+
+func (sl *virtualStreamListener) AcceptStream() (transport.StreamConn, error) {
+	sl.mu.Lock()
+	acceptCh := sl.acceptCh
+	sl.mu.Unlock()
+
+	select {
+	case acceptResponse, ok := <-acceptCh:
+		if !ok {
+			return nil, net.ErrClosed
+		}
+		return acceptResponse.conn, acceptResponse.err
+	case <-sl.closeCh:
+		return nil, net.ErrClosed
+	}
+}
+
+func (sl *virtualStreamListener) Close() error {
+	sl.mu.Lock()
+	defer sl.mu.Unlock()
+
+	if sl.acceptCh == nil {
+		return nil
+	}
+	sl.acceptCh = nil
+	close(sl.closeCh)
+	if sl.onCloseFunc != nil {
+		return sl.onCloseFunc()
+	}
+	return nil
+}
+
+func (sl *virtualStreamListener) Addr() net.Addr {
+	return sl.addr
+}
+
+type readRequest struct {
+	buffer []byte
+	respCh chan struct {
+		n    int
+		addr net.Addr
+		err  error
+	}
+}
+
+type virtualPacketConn struct {
+	net.PacketConn
+	mu          sync.Mutex // Mutex to protect access to the channels
+	readCh      chan readRequest
+	onCloseFunc OnCloseFunc
+}
+
+func (pc *virtualPacketConn) ReadFrom(p []byte) (n int, addr net.Addr, err error) {
+	respCh := make(chan struct {
+		n    int
+		addr net.Addr
+		err  error
+	}, 1)
+
+	pc.mu.Lock()
+	if pc.readCh == nil {
+		pc.mu.Unlock()
+		return 0, nil, net.ErrClosed
+	}
+	pc.readCh <- readRequest{
+		buffer: p,
+		respCh: respCh,
+	}
+	pc.mu.Unlock()
+
+	resp := <-respCh
+	return resp.n, resp.addr, resp.err
+}
+
+func (pc *virtualPacketConn) Close() error {
+	pc.mu.Lock()
+	defer pc.mu.Unlock()
+
+	if pc.readCh == nil {
+		return nil
+	}
+	pc.readCh = nil
+
+	if pc.onCloseFunc != nil {
+		return pc.onCloseFunc()
```

Comment:
You can also set the onCloseFun to nil. That way, there's no more reference to the function and the stuff it captures, even if a reference to the vpc lingers.

Same for the stream listener close.